src/pkg/packager/actions.go

```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package packager contains functions for interacting with, managing and deploying zarf packages.
+package packager
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/defenseunicorns/zarf/src/internal/packager/template"
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/exec"
+	"github.com/defenseunicorns/zarf/src/types"
+)
+
+// Run commands that a component has provided.
+func (p *Packager) runAction(actionSet types.ZarfComponentActionSet, actions []types.ZarfComponentAction, valueTemplate *template.Values) error {
+ACTION:
+	for _, a := range actions {
+		spinner := message.NewProgressSpinner("Running command \"%s\"", a.Cmd)
+		defer spinner.Success()
+
+		var (
+			ctx    context.Context
+			cancel context.CancelFunc
+			cmd    string
+			out    string
+			err    error
+			vars   map[string]string
+		)
+
+		// If the value template is not nil, get the variables for the action.
+		// No special variables or deprecations will be used the action.
+		// Reload the variables each time in case they have been changed by a previous action.
+		if valueTemplate != nil {
+			vars, _ = valueTemplate.GetVariables(types.ZarfComponent{})
+		}
+
+		cfg := actionGetCfg(actionSet, a, vars)
+
+		if cmd, err = actionCmdMutation(a.Cmd); err != nil {
+			spinner.Errorf(err, "Error mutating command: %s", cmd)
+		}
+
+		duration := time.Duration(cfg.MaxTotalSeconds) * time.Second
+		timeout := time.After(duration)
+
+		// Keep trying until the max retries is reached.
+		for remaining := cfg.MaxRetries + 1; remaining > 0; remaining-- {
+
+			// If no timeout is set, run the command and return.
+			if cfg.MaxTotalSeconds < 1 {
+				spinner.Updatef("Waiting for command \"%s\" (no timeout)", cmd)
+
+				// Try running the command and continue the retry loop if it fails.
+				if out, err = actionRun(context.TODO(), cfg, cmd); err != nil {
+					message.Debugf("command \"%s\" failed: %s", cmd, err.Error())
+					continue
+				}
+
+				// If an output variable is defined, set it.
+				if a.SetVariable != "" {
+					p.setVariable(a.SetVariable, out)
+				}
+
+				// If the command ran successfully, continue to the next action.
+				continue ACTION
```

Comment:
I'm pretty sure I'm missing something, but this looks like it will continue to outside of the original `for` loop, starting over from the beginning of the list of actions.

I'm just misunderstanding how continue's with labels work right?

Reply:
After a little googling it looks like I did misunderstand labels 🤷 


Would you be opposed to splitting `runAction()` into an additional function of `runActions()` that way we can return from a function instead of doing this continue-to-label thing?

Reply:
no that's fine, this just evolved into that and wasn't broken up better.